src/plugins/ipt/ipt.cpp

```diff
@@ -0,0 +1,321 @@
+    : pluginex(drakvuf, output)
+    , num_vcpus_{0}
+    , drakvuf_{drakvuf}
+{
+    if (!config.ipt_dir)
+    {
+        PRINT_DEBUG("[IPT] Target directory not provided, not activating IPT plugin\n");
+        return;
+    }
+
+    auto ipt_dir = std::filesystem::path(config.ipt_dir);
+    if (!std::filesystem::is_directory(ipt_dir))
+    {
+        PRINT_DEBUG("[IPT] Target directory doesn't exist. Creating...\n");
+        if (!std::filesystem::create_directory(ipt_dir))
+        {
+            PRINT_DEBUG("[IPT] Failed to create %s directory\n", ipt_dir.c_str());
+            throw -1;
+        }
+    }
+
+    {
+        auto vmi = vmi_lock_guard(drakvuf);
+        num_vcpus_ = vmi_get_num_vcpus(vmi);
+
+        // This is a DRAKVUF limitation
+        if (num_vcpus_ > MAX_DRAKVUF_VCPU)
+        {
+            PRINT_DEBUG("[IPT] Only first %d vCPUs will be traced\n", MAX_DRAKVUF_VCPU);
+            num_vcpus_ = MAX_DRAKVUF_VCPU;
+        }
+    }
+
+    // Always trace code branches, traces become kinda boring without them
+    // Ret compression may be sometimes problematic to reconstruct, disable it
+    uint64_t ipt_flags = DRAKVUF_IPT_BRANCH_EN | DRAKVUF_IPT_DIS_RETC;
+
+    if (config.trace_os)
+    {
+        PRINT_DEBUG("[IPT] Tracing OS\n");
+        ipt_flags |= DRAKVUF_IPT_TRACE_OS;
+    }
+    if (config.trace_user)
+    {
+        PRINT_DEBUG("[IPT] Tracing userspace\n");
+        ipt_flags |= DRAKVUF_IPT_TRACE_USR;
+    }
+
+    for (int i = 0; i < num_vcpus_; i++)
+    {
+        auto& vcpu = this->vcpus[i];
+        vcpu.id = i;
+        vcpu.last_offset = 0;
+        if (!drakvuf_enable_ipt(drakvuf, i, &vcpu.buf, &vcpu.size, ipt_flags))
+        {
+            PRINT_DEBUG("[IPT] Failed to enable IPT on vCPU %d\n", i);
+            throw -1;
+        }
+
+        std::stringstream ss;
+        ss << "ipt_stream_vcpu" << i;
+        auto stream_path = ipt_dir / ss.str();
+
+        vcpu.output_stream = std::ofstream(stream_path, std::ios::binary);
+        if (!vcpu.output_stream.is_open())
+        {
+            PRINT_DEBUG("Failed to open stream file for vCPU %d\n", i);
+            throw -1;
+        }
+    }
+
+
+    auto tr1 = register_trap(nullptr, &::ipt_cr3_cb,
+                             [](auto drakvuf, auto info, auto trap) -> drakvuf_trap_t*
+    {
+        trap->type = REGISTER;
+        trap->reg = CR3;
+
+        if (!drakvuf_add_trap(drakvuf, trap))
+            return nullptr;
+        return trap;
+    }, "ipt_cr3", UNLIMITED_TTL);
+
+    if (!tr1)
+    {
+        PRINT_DEBUG("[IPT] Failed to register CR3 trap");
+        throw -1;
+    }
+
+    auto tr2 = register_trap(nullptr, &::ipt_catchall_cb,
+                             [](auto drakvuf, auto info, auto trap) -> drakvuf_trap_t*
+    {
+        trap->type = CATCHALL_BREAKPOINT;
+        if (!drakvuf_add_trap(drakvuf, trap))
+            return nullptr;
+        return trap;
+    }, "ipt_catchall", UNLIMITED_TTL);
```

Comment:
This is kinda unreadable even for me. Replace lambdas with regular functions to cleanup.